File: bin/cellbarcode_filter.py
#!/usr/bin/env python3
"""
CellBarcode-style filtering for NextClone discovery mode

Implements filtering strategies from the CellBarcode paper:
1. Automatic threshold filtering (k-means clustering on log-transformed counts)
2. Cluster filtering (remove barcodes similar to more abundant ones)
3. UMI filtering (if applicable)

Reference: Sun et al. (2024) Nature Computational Science
"""

import sys
import argparse
import numpy as np
from collections import Counter
from typing import List, Tuple, Dict


def auto_threshold_filter(barcode_counts: Dict[str, int], min_count: int = 1) -> Tuple[List[str], int]:
    """
    Automatic threshold filtering using 1D k-means clustering.
    
    Algorithm from CellBarcode:
    1. Remove barcodes with count below median
    2. Transform counts by log2(x+1)
    3. Apply 1D k-means clustering with k=2
    4. Use boundary between clusters as threshold
    
    Args:
        barcode_counts: Dict mapping barcode -> count
        min_count: Minimum count to consider
    
    Returns:
        Tuple of (filtered_barcodes, threshold_used)
    """
    # Filter out very low count barcodes
    filtered = {bc: count for bc, count in barcode_counts.items() if count >= min_count}
    
    if len(filtered) < 2:
        return list(filtered.keys()), min_count
    
    # Get counts and compute median
    counts = np.array(list(filtered.values()))
    median_count = np.median(counts)
    
    # Remove barcodes below median
    above_median = {bc: count for bc, count in filtered.items() if count >= median_count}
    
    if len(above_median) < 2:
        return list(filtered.keys()), min_count
    
    # Log transform
    log_counts = np.log2(np.array(list(above_median.values())) + 1)
    
    # 1D k-means with k=2
    # Initialize with min and max
    centers = np.array([log_counts.min(), log_counts.max()])
    
    # Iterate until convergence
    for _ in range(100):
        # Assign to nearest center
        assignments = np.argmin(np.abs(log_counts[:, None] - centers[None, :]), axis=1)
        
        # Update centers
        new_centers = np.array([
            log_counts[assignments == 0].mean() if np.any(assignments == 0) else centers[0],
            log_counts[assignments == 1].mean() if np.any(assignments == 1) else centers[1]
        ])
        
        # Check convergence
        if np.allclose(centers, new_centers):
            break
        centers = new_centers
    
    # Threshold is the midpoint between centers
    threshold_log = (centers[0] + centers[1]) / 2
    threshold = int(2 ** threshold_log - 1)
    
    # Filter barcodes
    filtered_barcodes = [bc for bc, count in filtered.items() if count >= threshold]
    
    return filtered_barcodes, threshold


def hamming_distance(s1: str, s2: str) -> int:
    """Calculate Hamming distance between two strings of equal length."""
    if len(s1) != len(s2):
        raise ValueError("Strings must have equal length")
    return sum(c1 != c2 for c1, c2 in zip(s1, s2))


def cluster_filter(barcode_counts: Dict[str, int], max_distance: int = 1, 
                   distance_metric: str = 'hamming') -> List[str]:
    """
    Cluster filtering: remove barcodes with small edit distance to more abundant ones.
    
    Algorithm:
    1. Sort barcodes by abundance (descending)
    2. For each barcode, compare to all more abundant barcodes
    3. If distance < threshold to any more abundant barcode, remove it
    
    Args:
        barcode_counts: Dict mapping barcode -> count
        max_distance: Maximum edit distance to consider as cluster member
        distance_metric: 'hamming' or 'levenshtein'
    
    Returns:
        List of filtered barcodes (true barcodes)
    """
    # Sort by count descending
    sorted_barcodes = sorted(barcode_counts.items(), key=lambda x: -x[1])
    
    kept_barcodes = []
    
    for barcode, count in sorted_barcodes:
        # Check if this barcode is similar to any already-kept barcode
        is_cluster_member = False
        
        for kept_bc in kept_barcodes:
            if distance_metric == 'hamming':
                if len(barcode) == len(kept_bc):
                    dist = hamming_distance(barcode, kept_bc)
                else:
                    # Different lengths, skip hamming check
                    continue
            else:
                # For now, use hamming only
                # Could implement levenshtein if needed
                continue
            
            if dist <= max_distance:
                is_cluster_member = True
                break
        
        if not is_cluster_member:
            kept_barcodes.append(barcode)
    
    return kept_barcodes


def manual_threshold_filter(barcode_counts: Dict[str, int], threshold: int) -> List[str]:
    """
    Manual threshold filtering: keep barcodes with count >= threshold.
    
    Args:
        barcode_counts: Dict mapping barcode -> count
        threshold: Minimum count to keep
    
    Returns:
        List of filtered barcodes
    """
    return [bc for bc, count in barcode_counts.items() if count >= threshold]


def apply_filtering(barcode_counts_file: str, output_file: str, 
                    filter_method: str = 'auto', threshold: int | None = None,
                    cluster_distance: int = 1, min_count: int = 1) -> None:
    """
    Apply CellBarcode-style filtering to discovered barcodes.
    
    Args:
        barcode_counts_file: Input file with barcode\\tcount format
        output_file: Output file for filtered barcodes
        filter_method: 'auto', 'manual', 'cluster', or 'combined'
        threshold: Manual threshold (required if filter_method='manual')
        cluster_distance: Max edit distance for cluster filtering
        min_count: Minimum count to consider
    """
    # Read barcode counts
    barcode_counts = {}
    with open(barcode_counts_file, 'r') as f:
        for line in f:
            parts = line.strip().split('\t')
            if len(parts) == 2:
                barcode, count = parts
                barcode_counts[barcode] = int(count)
    
    print(f"Read {len(barcode_counts)} barcodes from {barcode_counts_file}", file=sys.stderr)
    
    # Apply filtering
    if filter_method == 'auto':
        filtered_barcodes, threshold_used = auto_threshold_filter(barcode_counts, min_count)
        print(f"Auto threshold: {threshold_used}", file=sys.stderr)
    
    elif filter_method == 'manual':
        if threshold is None:
            raise ValueError("Manual threshold must be specified")
        filtered_barcodes = manual_threshold_filter(barcode_counts, threshold)
        print(f"Manual threshold: {threshold}", file=sys.stderr)
    
    elif filter_method == 'cluster':
        filtered_barcodes = cluster_filter(barcode_counts, cluster_distance)
        print(f"Cluster filtering with max distance: {cluster_distance}", file=sys.stderr)
    
    elif filter_method == 'combined':
        # First apply auto threshold, then cluster filtering
        threshold_barcodes, threshold_used = auto_threshold_filter(barcode_counts, min_count)
        threshold_counts = {bc: barcode_counts[bc] for bc in threshold_barcodes}
        filtered_barcodes = cluster_filter(threshold_counts, cluster_distance)
        print(f"Combined: auto threshold {threshold_used}, then cluster distance {cluster_distance}", 
              file=sys.stderr)
    
    else:
        raise ValueError(f"Unknown filter method: {filter_method}")
    
    # Write output
    with open(output_file, 'w') as f:
        for barcode in filtered_barcodes:
            f.write(f"{barcode}\n")
    
    print(f"Filtered to {len(filtered_barcodes)} barcodes", file=sys.stderr)
    print(f"Output written to {output_file}", file=sys.stderr)


def main():
    parser = argparse.ArgumentParser(
        description='CellBarcode-style filtering for NextClone discovery mode',
        formatter_class=argparse.ArgumentDefaultsHelpFormatter
    )
    
    parser.add_argument('input', help='Input barcode counts file (barcode\\tcount)')
    parser.add_argument('output', help='Output filtered barcodes file')
    parser.add_argument('--method', choices=['auto', 'manual', 'cluster', 'combined'],
                       default='auto', help='Filtering method')
    parser.add_argument('--threshold', type=int, help='Manual threshold (for method=manual)')
    parser.add_argument('--cluster-distance', type=int, default=1,
                       help='Max edit distance for cluster filtering')
    parser.add_argument('--min-count', type=int, default=1,
                       help='Minimum count to consider')
    
    args = parser.parse_args()
    
    apply_filtering(
        args.input, args.output,
        filter_method=args.method,
        threshold=args.threshold,
        cluster_distance=args.cluster_distance,
        min_count=args.min_count
    )


if __name__ == '__main__':
    main()

File: bin/cellbarcode_filter.R
#!/usr/bin/env Rscript
# CellBarcode-based filtering for NextClone discovery mode
# Uses the official CellBarcode R package for barcode filtering
#
# Reference: Sun et al. (2024) Nature Computational Science
# Package: https://github.com/wenjie1991/CellBarcode

suppressPackageStartupMessages({
  library(CellBarcode)
  library(data.table)
})

# Parse command line arguments
args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 2) {
  stop("Usage: cellbarcode_filter.R <input_counts_file> <output_filtered_file> [method] [threshold] [cluster_distance] [min_count]")
}

input_file <- args[1]
output_file <- args[2]
method <- ifelse(length(args) >= 3, args[3], "auto")
threshold <- ifelse(length(args) >= 4, as.numeric(args[4]), NA)
cluster_distance <- ifelse(length(args) >= 5, as.numeric(args[5]), 1)
min_count <- ifelse(length(args) >= 6, as.numeric(args[6]), 1)

# Read barcode counts
message("Reading barcode counts from ", input_file)
counts_data <- fread(input_file, header = FALSE, sep = "\t")
setnames(counts_data, c("barcode_seq", "count"))

message("Read ", nrow(counts_data), " barcodes")

# Filter by minimum count first
counts_data <- counts_data[count >= min_count]
message("After min_count filter (", min_count, "): ", nrow(counts_data), " barcodes")

if (nrow(counts_data) == 0) {
  stop("No barcodes remaining after min_count filter")
}

# Create BarcodeObj from the count data
# CellBarcode expects a matrix with barcodes as rows and samples as columns
# We have a single sample, so create a matrix with one column
barcode_matrix <- as.matrix(counts_data$count)
rownames(barcode_matrix) <- counts_data$barcode_seq
colnames(barcode_matrix) <- "sample1"

# Create BarcodeObj
bc_obj <- bc_create_BarcodeObj(barcode_matrix, sample_name = "sample1")

message("Created BarcodeObj with ", length(bc_barcodes(bc_obj)), " barcodes")

# Apply filtering based on method
if (method == "auto") {
  # Automatic threshold using CellBarcode's bc_auto_cutoff
  message("Applying automatic threshold filtering (k-means)")
  
  # bc_cure_depth with depth=-1 uses automatic cutoff
  bc_obj <- bc_cure_depth(bc_obj, depth = -1)
  
  # Get the cutoff that was used
  cutoff <- bc_auto_cutoff(bc_obj, useCleanBc = FALSE)
  message("Automatic threshold: ", round(cutoff, 2))
  
} else if (method == "manual") {
  # Manual threshold
  if (is.na(threshold)) {
    stop("Manual method requires threshold parameter")
  }
  
  message("Applying manual threshold: ", threshold)
  bc_obj <- bc_cure_depth(bc_obj, depth = threshold)
  
} else if (method == "cluster") {
  # Cluster filtering using CellBarcode's bc_cure_cluster
  message("Applying cluster filtering (max distance: ", cluster_distance, ")")
  
  # First need to cure depth (use min_count as threshold)
  bc_obj <- bc_cure_depth(bc_obj, depth = min_count)
  
  # Then cluster similar barcodes
  bc_obj <- bc_cure_cluster(
    bc_obj,
    dist_threshold = cluster_distance,
    dist_method = "hamm"  # Hamming distance
  )
  
} else if (method == "combined") {
  # Combined: auto threshold + cluster filtering
  message("Applying combined filtering (auto + cluster)")
  
  # Step 1: auto threshold
  bc_obj <- bc_cure_depth(bc_obj, depth = -1)
  cutoff <- bc_auto_cutoff(bc_obj, useCleanBc = FALSE)
  message("Step 1 - Auto threshold: ", round(cutoff, 2))
  
  # Step 2: cluster filtering
  message("Step 2 - Cluster filtering (max distance: ", cluster_distance, ")")
  bc_obj <- bc_cure_cluster(
    bc_obj,
    dist_threshold = cluster_distance,
    dist_method = "hamm"
  )
  
} else {
  stop("Unknown method: ", method, ". Use 'auto', 'manual', 'cluster', or 'combined'")
}

# Extract filtered barcodes
filtered_barcodes <- bc_barcodes(bc_obj, unlist = TRUE)
message("Filtered to ", length(filtered_barcodes), " barcodes")

# Write output (just the barcode sequences, one per line)
writeLines(filtered_barcodes, output_file)
message("Output written to ", output_file)
